src/client/dfs/duns.c

```diff
@@ -373,39 +424,93 @@ duns_resolve_path(const char *path, struct duns_attr_t *attr)
 		return err;
 	}
 
+	realp = realpath(path, NULL);
+	if (realp == NULL)
+		return errno;
+
+	path_len = strnlen(realp, PATH_MAX);
+	if (path_len > PATH_MAX - 1)
+		D_GOTO(out, rc = ENAMETOOLONG);
+
+	D_ALLOC(rel_path, path_len + 1);
+	if (rel_path == NULL)
+		D_GOTO(out, rc = ENOMEM);
+
+	D_STRDUP(dir_path, realp);
+	if (dir_path == NULL)
+		D_GOTO(out, rc = ENOMEM);
+
+	cur_idx = path_len;
+
+	while (1) {
 #ifdef LUSTRE_INCLUDE
-	if (fs.f_type == LL_SUPER_MAGIC) {
-		rc = duns_resolve_lustre_path(path, attr);
-		if (rc == 0)
-			return 0;
+		if (fs.f_type == LL_SUPER_MAGIC) {
+			rc = duns_resolve_lustre_path(dir_path, attr);
+			if (rc == 0)
+				D_GOTO(out, rc);
 
-		/* if Lustre specific method fails, fallback to try
-		 * the normal way...
-		 */
-	}
+			/* if Lustre specific method fails, fallback to try the
+			 * normal way...
+			 */
+		}
 #endif
 
-	s = lgetxattr(path, DUNS_XATTR_NAME, &str, DUNS_MAX_XATTR_LEN);
-	if (s < 0 || s > DUNS_MAX_XATTR_LEN) {
-		int err = errno;
+		s = lgetxattr(dir_path, DUNS_XATTR_NAME, &str,
+			      DUNS_MAX_XATTR_LEN);
+		if (s < 0 || s > DUNS_MAX_XATTR_LEN) {
+			int err = errno;
 
-		if (err == ENOTSUP) {
-			D_INFO("Path is not in a filesystem that supports the"
-				" DAOS unified namespace\n");
-		} else if (err == ENODATA) {
-			D_INFO("Path does not represent a DAOS link\n");
-		} else if (s > DUNS_MAX_XATTR_LEN) {
-			err = EIO;
-			D_ERROR("Invalid xattr length\n");
-		} else {
-			D_ERROR("Invalid DAOS unified namespace xattr: %s\n",
-				strerror(err));
+			if (err == ENODATA) {
+				if (cur_idx == 0 || attr->da_no_reverse_lookup)
+					D_INFO("Path does not represent a DAOS"
+					       " link\n");
+				else
+					goto parse;
+			} else if (err == ENOTSUP) {
+				D_INFO("Path is not in a filesystem that"
+				       " supports the DAOS unified namespace\n");
+			} else if (s > DUNS_MAX_XATTR_LEN) {
+				err = EIO;
+				D_ERROR("Invalid xattr length\n");
+			} else {
+				D_ERROR("Invalid DAOS unified namespace xattr:"
+					" %s\n", strerror(err));
+			}
+
+			D_GOTO(out, rc = err);
 		}
 
-		return err;
+		/** On success, parse the attribute */
+		rc = duns_parse_attr(&str[0], s, attr);
+		if (rc) {
+			D_ERROR("Invalid xattr format\n");
+			D_GOTO(out, rc);
+		}
+		/** if the xattr parsing succeeds, break */
+		break;
+parse:
+		rc = parse_path(realp, path_len, &cur_idx, &rel_len, dir_path,
+				rel_path);
+		if (rc) {
+			D_ERROR("Failed to parse %s (%s)\n",
+				path, strerror(rc));
+			D_GOTO(out, rc);
+		}
 	}
 
-	return duns_parse_attr(&str[0], s, attr);
+	if (cur_idx != path_len) {
+		D_ASSERT(rel_path);
+		attr->da_rel_path = strdup(rel_path);
```

Comment:
Should we use `strndup`?
```suggestion
		attr->da_rel_path = strndup(rel_path, rel_len);
```

Reply:
i don't think it's necessary because we already checked the length above of the input path.
strndup would be required if this was on an input string that we don't know the length of.